castle/cms/cron/_auto_publish_retract.py

```diff
@@ -0,0 +1,74 @@
+from AccessControl.SecurityManagement import newSecurityManager
+from DateTime import DateTime
+from plone import api
+from Products.CMFPlone.interfaces.siteroot import IPloneSiteRoot
+from tendo import singleton
+from zope.component.hooks import setSite
+
+import logging
+import transaction
+
+
+logger = logging.getLogger('castle.cms')
+
+
+def run_query(query_params):
+
+    catalog = api.portal.get_tool('portal_catalog')
+    content = catalog(**query_params)
+    
+    to_state = 'published' if query_params.get('effective') else 'private'
+    for brain in content:
+        obj = brain.getObject()
+        api.content.transition(
+            obj=obj, 
+            to_state=to_state
+        )
+        obj.setModificationDate()
+        obj.reindexObject(idxs=['modified'])
+    transaction.commit()
+
+
+def set_queries(site):
+    setSite(site)
+
+    # publish pending content with an effective date within 30 minutes of cron run
+    start = DateTime() - (30.0 / (24 * 60))
```

Comment:
I'm not sure I follow the `30.0 / (24 * 60)` bit -- what value does represent?

I apologize if this is obvious, I'm probably just be dense. I would expect "30 minutes ago" to be something more like `DateTime() - (30.0 * 60)`

or maybe a python standard library take on it (eschewing the zope DateTime), like:

```
datetime.timestamp(datetime.now() - timedelta(minutes=30))
```

unless there's some hidden timezone thing that happens with the zope DateTime (I don't know off hand)

Reply:
No worries, it's definitely strange. This is the syntax I found to subtract 30 minutes from a Zope DateTime object. The math evaluates to '0.02083333', which is the exact value needed to subtract 30 minutes from the object. I have no idea how this resembles 30 minutes, but it makes sense to Zope. Here are the start/end dates when running:

`
2024/01/08 07:37:49.145811 HST
2024/01/08 08:07:49.145848 HST
`

I could replace the math with a hardcoded float, though I think it would make it more confusing.

I could also convert it to a standard python datetime object for legibility and then use timedelta, though I think it would need to be converted back to Zope for the query.

As for timezone, I found DateTime() utilized similarly in other cron scripts without any timezone adjustments, so I assume Zope does something under the hood for that.

Reply:
is there a link to the discussion where you found this?

Reply:
I couldn't find the discussion, but after digging into the DateTime module I found a method that converts it to epoch time, so now it just subtracts 30 minutes from epoch and re-converts it into DateTime, much more sensible logic. My fault for not finding that method earlier.

Reply:
awesome, thank you!